core/src/main/java/org/elasticsearch/index/engine/IndexSearcherWrappingService.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.search.IndexSearcher;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.index.engine.Engine.Searcher;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Service responsible for wrapping the {@link DirectoryReader} of a {@link Searcher} with the
+ * configured {@link IndexSearcherWrapper} instances.
+ */
+public final class IndexSearcherWrappingService {
+
+    private final Set<IndexSearcherWrapper> wrappers;
+
+    // for unit tests:
+    IndexSearcherWrappingService() {
+        this.wrappers = new HashSet<>();
+    }
+
+    @Inject
+    public IndexSearcherWrappingService(Set<IndexSearcherWrapper> wrappers) {
```

Comment:
> Why a Set? 

This gets injected via Guice and there is no list binder, just a set binder. We can add `IndexSearcherWrapper#order()` that returns an integer and change the incoming Set to a list and sort based on the order?

>  Can't someone who wants to add multiple layers of IndexSearcher wrappers do it with a single IndexSearcherWrapper?

They could, but if multiple plugins like to independently decorate the IndexSearcher then that wouldn't possible if only a single instance can be specified. I don't have a concrete example... but I like to be flexible around this.